azkaban-common/src/main/java/azkaban/executor/ExecutorHealthChecker.java

```diff
@@ -124,11 +119,26 @@ public void checkExecutorHealth() {
           }
         }
       } catch (final ExecutorManagerException e) {
-        handleExecutorNotAliveCase(entry, executor, e);
+        handleExecutorNotAliveCase(executor, entry.getValue(), e);
       }
     }
   }
 
+  /**
+   * Finalize given flows with the provided reason.
+   *
+   * @param flows
+   * @param finalizeReason
+   */
+  private void finalizeFlows(List<ExecutableFlow> flows, String finalizeReason) {
+    for (ExecutableFlow flow: flows) {
+      logger.warn(
```

Comment:
Does it make sense to put the body of the loop in a try-catch in order to continue the loop in case of a runtime exception.

Reply:
Good point.  I thought about exception handling for this. The underlying utility function `finalizeFlow` already captures any `ExecutorManagerExceptions` (which is not a RuntimeException). That is the most likely source of exceptions over here and with that already handled it made sense to pass on any unexpected RuntimeExceptions to the callers as soon as possible rather than swallowing it (even if logged).

Reply:
Not catching a Runtime exception means that we will skip finalizing any subsequent flows. That's the decision you have to make - whether to "miss" finalizing of subsequent Flow or not.

Reply:
Agree that's the current behavior. However making such a change would mean either:
(1) Any unreported exceptions (including significant ones like NPEs) will at the most simply be logged and things will continue business as usual. (This is part of the reason why we see an increased exceptions in our logs without having much in the way of tracking them)
(2) Alternatively such exceptions could be accumulated/consolidated and thrown to a higher layer but that has additional implementation and maintenance costs.
In this specific case both, the possibility of RuntimeExceptions at this point, and  the drawback of not finalizing subsequent flows seemed low enough that throwing the exception immediately made sense.

Reply:
What will be the impact of getting a Runtime exception when the flow is being finalized:
- One, is that the flow and any subsequent flows yet to be iterated upon will not get finalized.
- Will some thread get killed or a JVM exit? (basically where will this exception be caught, because that's where things will get impacted)

That's the analysis we will have to weigh.

Also - you are bringing up a very good point (Thanks)

> (This is part of the reason why we see an increased exceptions in our logs without having much in the way of tracking them)

We should assume that people do monitor and get alerted to exception being printed in the log. I am trying to get this addressed (LIHADOOP-52504) internally for us.

Reply:
While reviewing existing code for this discovered larger issues with exception handling here. Creating a separate ticket for it and will address this as part of a separate PR.

Reply:
Thanks. That was the purpose of my pushing you to do this exercise. It's ok to address this in a follow-on. Thanks again for the follow up. I appreciate it.